src/Catalog/CatalogCommitUtilities.cs

```diff
@@ -0,0 +1,457 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using NuGet.Common;
+
+namespace NuGet.Services.Metadata.Catalog
+{
+    public static class CatalogCommitUtilities
+    {
+        /// <summary>
+        /// Creates an enumerable of <see cref="CatalogCommitItemBatch" /> instances.
+        /// </summary>
+        /// <remarks>
+        /// A <see cref="CatalogCommitItemBatch" /> instance contains only the latest commit for each package identity.
+        /// </remarks>
+        /// <param name="catalogItems">An enumerable of <see cref="CatalogCommitItem" />.</param>
+        /// <param name="getCatalogCommitItemKey">A function that returns a key for a <see cref="CatalogCommitItem" />.</param>
+        /// <returns>An enumerable of <see cref="CatalogCommitItemBatch" />.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="catalogItems" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="getCatalogCommitItemKey" /> is <c>null</c>.</exception>
+        public static IEnumerable<CatalogCommitItemBatch> CreateCommitItemBatches(
+            IEnumerable<CatalogCommitItem> catalogItems,
+            GetCatalogCommitItemKey getCatalogCommitItemKey)
+        {
+            if (catalogItems == null)
+            {
+                throw new ArgumentNullException(nameof(catalogItems));
+            }
+
+            if (getCatalogCommitItemKey == null)
+            {
+                throw new ArgumentNullException(nameof(getCatalogCommitItemKey));
+            }
+
+            var catalogItemsGroups = catalogItems
+                .GroupBy(catalogItem => getCatalogCommitItemKey(catalogItem));
+
+            foreach (var catalogItemsGroup in catalogItemsGroups)
+            {
+                var key = catalogItemsGroup.Key;
+
+                // Before filtering out all but the latest commit for each package identity, determine the earliest
+                // commit timestamp for all items in this batch.  This timestamp is important for processing commits
```

Comment:
>timestamp [](start = 71, length = 9)

I don't understand why the min *before filtering* is important. Let's talk offline if it's complex.

Reply:
In my V3 performance work I have striven to preserve existing assumptions and behaviors as much as possible, no matter how small or obscure.  Before my commit parallelization, Catalog2Registration would update its cursor only when all commits up to the present commit were fully and successfully processed.  I aimed to preserve that behavior even with commit parallelization.  The minimum commit timestamp before filtering enables this.

I'm open to changing the behavior to completely ignore older commits and their timestamps, but I think it should be separate from this PR.